samza-azure/src/main/java/org/apache/samza/AzureJobCoordinator.java

```diff
@@ -0,0 +1,483 @@
+    streamMetadataCache = StreamMetadataCache.apply(METADATA_CACHE_TTL_MS, config);
+    table.addProcessorEntity(INITIAL_STATE, processorId, azureLeaderElector.amILeader());
+
+    // Start scheduler for heartbeating
+    LOG.info("Starting scheduler for heartbeating.");
+    heartbeat.scheduleTask();
+
+    azureLeaderElector.tryBecomeLeader();
+
+    // Start scheduler to check for job model version upgrades
+    LOG.info("Starting scheduler to check for job model version upgrades.");
+    versionUpgrade.setStateChangeListener(createJMVersionUpgradeListener());
+    versionUpgrade.scheduleTask();
+
+    // Start scheduler to check for leader liveness
+    LOG.info("Starting scheduler to check for leader liveness.");
+    leaderAlive.setStateChangeListener(createLeaderLivenessListener());
+    leaderAlive.scheduleTask();
+  }
+
+  @Override
+  public void stop() {
+    LOG.info("Shutting down Azure job coordinator.");
+
+    if (coordinatorListener != null) {
+      coordinatorListener.onJobModelExpired();
+    }
+
+    // Resign leadership
+    if (azureLeaderElector.amILeader()) {
+      azureLeaderElector.resignLeadership();
+    }
+
+    // Shutdown all schedulers
+    shutdownSchedulers();
+
+    if (coordinatorListener != null) {
+      coordinatorListener.onCoordinatorStop();
+    }
+  }
+
+  @Override
+  public String getProcessorId() {
+    return processorId;
+  }
+
+  @Override
+  public void setListener(JobCoordinatorListener listener) {
+    this.coordinatorListener = listener;
+  }
+
+  @Override
+  public JobModel getJobModel() {
+    return jobModel;
+  }
+
+  private void shutdownSchedulers() {
+    heartbeat.shutdown();
+    leaderAlive.shutdown();
+    liveness.shutdown();
+    renewLease.shutdown();
+    if (leaderBarrierScheduler != null) {
+      leaderBarrierScheduler.shutdown();
+    }
+    versionUpgrade.shutdown();
+  }
+
+  /**
+   * Creates a listener for LeaderBarrierCompleteScheduler class.
+   * Invoked by the leader when it detects that rebalancing has completed by polling the processor table.
+   * Updates the barrier state on the blob to denote that the barrier has completed.
+   * Cancels all future tasks scheduled by the LeaderBarrierComplete scheduler to check if barrier has completed.
+   * @return an instance of SchedulerStateChangeListener.
+   */
+  private SchedulerStateChangeListener createLeaderBarrierCompleteListener(String nextJMVersion, AtomicBoolean barrierTimeout) {
+    return () -> {
+      versionUpgradeDetected.getAndSet(false);
+      String state;
+      if (barrierTimeout.get()) {
+        LOG.error("Barrier timed out for version {}", nextJMVersion);
+        state = BarrierState.TIMEOUT.name() + " " + nextJMVersion;
+      } else {
+        LOG.info("Leader detected barrier completion.");
+        state = BarrierState.END.name() + " " + nextJMVersion;
+      }
+      if (!leaderBlob.publishBarrierState(state, azureLeaderElector.getLeaseId().get())) {
```

Comment:
Seems that this can only be performed by the leader. What if during the wait for barrier completion, leader has lost its lease to the leaderBlob? I did not see a logic here to handle that case.